read-fonts/src/tables/gsub/closure.rs

```diff
@@ -241,247 +368,475 @@ impl GlyphClosure for LigatureSubstFormat1<'_> {
-                            .map(|rec| rec.lookup_list_index()),
-                    )
+                if !rule.matches_classes(&our_classes) {
+                    continue;
+                }
+
+                let mut seen_sequence_indices = IntSet::new();
+                for lookup_record in rule.lookup_records() {
+                    let lookup_id = lookup_record.lookup_list_index();
+                    let seq_idx = lookup_record.sequence_index();
+                    let active_glyphs = if !seen_sequence_indices.insert(seq_idx) {
+                        None
+                    } else if seq_idx == 0 {
+                        Some(intersect_class(&classdef, &cur_glyphs, class_i))
+                    } else {
+                        Some(intersect_class(
+                            &classdef,
+                            ctx.glyphs(),
+                            rule.input_sequence()[seq_idx as usize - 1].get(),
+                        ))
+                    };
+
+                    ctx.add_todo(lookup_id, active_glyphs);
                 }
             }
         }
         Ok(())
     }
 }
 
-impl ChainedSequenceContextFormat3<'_> {
-    fn add_reachable_lookups(
-        &self,
-        glyphs: &IntSet<GlyphId16>,
-        lookups: &mut IntSet<u16>,
-    ) -> Result<(), ReadError> {
-        for coverage in self
-            .backtrack_coverages()
+// these are basically the same; but we need to jump through some hoops
+// to get the fields to line up
+enum ContextFormat3<'a> {
+    Plain(SequenceContextFormat3<'a>),
+    Chain(ChainedSequenceContextFormat3<'a>),
+}
+
+impl ContextFormat3<'_> {
+    fn coverages(&self) -> ArrayOfOffsets<CoverageTable> {
+        match self {
+            ContextFormat3::Plain(table) => table.coverages(),
+            ContextFormat3::Chain(table) => table.input_coverages(),
+        }
+    }
+
+    fn lookup_records(&self) -> &[SequenceLookupRecord] {
+        match self {
+            ContextFormat3::Plain(table) => table.seq_lookup_records(),
+            ContextFormat3::Chain(table) => table.seq_lookup_records(),
+        }
+    }
+
+    fn matches_glyphs(&self, glyphs: &IntSet<GlyphId16>) -> bool {
+        let (backtrack, lookahead) = match self {
+            Self::Plain(_) => (None, None),
+            Self::Chain(table) => (
+                Some(table.backtrack_coverages()),
+                Some(table.lookahead_coverages()),
+            ),
+        };
+        self.coverages()
             .iter()
-            .chain(self.input_coverages().iter())
-            .chain(self.lookahead_coverages().iter())
-        {
-            if !coverage?.iter().any(|gid| glyphs.contains(gid)) {
-                return Ok(());
-            }
+            .chain(backtrack.into_iter().flat_map(|x| x.iter()))
+            .chain(lookahead.into_iter().flat_map(|x| x.iter()))
+            .all(|cov| {
+                cov.map(|cov| cov.iter().any(|gid| glyphs.contains(gid)))
```

Comment:
Suggest to use this API to check if a coverage table intersects a glyph set: https://github.com/googlefonts/fontations/blob/ae80c78ff4a886539e66306962a8a7cb71505e1e/read-fonts/src/tables/layout.rs#L199

Reply:
hmm this takes `GlyphId` but the current code uses `GlyphId16`, I'll leave this for you to resolve.